auacm/app/static/js/controllers/ScoreboardController.js

```diff
@@ -61,38 +13,95 @@ app.controller('ScoreboardController', ['$scope', '$http', '$routeParams', '$win
-                $scope.timeLeft = ($scope.competition.startTime + $scope.competition.length - thisTime) / 1000;
-                thisTime = new Date().getTime();
-                if ($scope.timeLeft <= 0) {
-                    $scope.timeLeft = 0;
-                    $interval.cancel(Timer);
-                }
-            }, 1000);
+            var thisTime = Math.floor(new Date().getTime() / 1000);
+            if (thisTime < $scope.competition.startTime + $scope.competition.length) {
+                // if the competition is still active
+                $scope.active = true;
+                $scope.timeLeft = $scope.competition.startTime + $scope.competition.length
+                        - thisTime;
+                var timer = $interval(function() {
+                    $scope.timeLeft = $scope.competition.startTime + $scope.competition.length 
+                            - thisTime;
+                    thisTime = Math.floor(new Date().getTime() / 1000);
+                    if ($scope.timeLeft <= 0) {
+                        $scope.active = false;
+                        $scope.timeLeft = 0;
+                        $interval.cancel(timer);
+                    }
+                }, 1000);
+            }
+            
         },
         function(error) {
             
         });
+}]);
 
-
-    // Open the socket connection
-    var socket = io.connect('http://' + $window.location.host + '/judge')
-    socket.on('connect', function() {
-        console.log('connected');
+var genScoreboard = function($scope) {
+    for (var i = 0; i < $scope.teams.length; i++) {
+        var team = $scope.teams[i];
+        var solved = 0;
+        var time = 0;
+        for (var problem in team.problemData) {
+            if (team.problemData[problem].status == 'correct') {
+                solved++;
+                time += team.problemData[problem].problemTime;
+            }
+        }
+        team.solved = solved;
+        team.time = time;
+    }
+    
+    $scope.teams.sort(function(a, b) {
+        if (a.solved != b.solved) {
+            return b.solved - a.solved;
+        } else {
+            return a.time - b.time;
+        }
     });
+    
+    var rank = 1;
+    var prevSolved = $scope.teams[0].solved;
+    var prevTime = $scope.teams[0].solved;
+    $scope.teams[0].rank = rank;
+    for (var i = 1; i < $scope.teams.length; i++) {
+        var team = $scope.teams[i];
+        if (team.solved < prevSolved) {
+            rank++;
+            team.rank = rank;
+        } else if (team.solved == prevSolved && team.time > prevTime) {
+            rank++;
+            team.rank = rank;
+        } else {
+            team.rank = rank;
+        }
+        prevSolved = team.solved;
+        prevTime = team.time;
+    }
+
+    // Can break if an $apply is already in progress, need to check first (???)
+    if (!$scope.$$phase) {
+        $scope.$digest();
+    }
+}
 
+var connectToSocket = function($window, $scope) {
```

Comment:
Since it's outside the controller, doesn't it have global scope? That might cause a problem if we want to create a ```connectToSocket()``` function elsewhere.